substratevm/src/com.oracle.svm.jni/src/com/oracle/svm/jni/access/JNIAccessFeature.java

```diff
@@ -184,28 +184,48 @@ private static ConditionalConfigurationRegistry getConditionalConfigurationRegis
-        ResolvedJavaField field = JNIAccessibleMethod.getCallWrapperField(wrappedMetaAccess, variant, nonVirtual);
+    private void createJavaCallTrampoline(BeforeAnalysisAccessImpl access, CallVariant variant, boolean nonVirtual) {
+        AnalysisMetaAccess metaAccess = access.getMetaAccess();
+        ResolvedJavaField field = JNIAccessibleMethod.getCallWrapperField(metaAccess.getWrapped(), variant, nonVirtual);
         access.getUniverse().lookup(field.getDeclaringClass()).registerAsReachable();
         access.registerAsAccessed(access.getUniverse().lookup(field));
-        ResolvedJavaMethod method = JNIJavaCallWrappers.lookupJavaCallTrampoline(wrappedMetaAccess, variant, nonVirtual);
-        JNICallTrampolineMethod trampoline = new JNICallTrampolineMethod(method, field, nonVirtual);
+        String trampolineName = JNIJavaCallTrampolines.getTrampolineName(variant, nonVirtual);
+        Method reflectionMethod = ReflectionUtil.lookupMethod(JNIJavaCallTrampolines.class, trampolineName);
+        // Look up the java method to ensure a JNICallTrampolineMethod gets created for it through
+        // com.oracle.svm.jni.hosted.JNINativeCallWrapperSubstitutionProcessor.lookup
+        metaAccess.lookupJavaMethod(reflectionMethod);
+        JNICallTrampolineMethod trampoline = getCallTrampolineMethod(trampolineName);
         access.registerAsCompiled(access.getUniverse().lookup(trampoline));
-        return trampoline;
     }
 
     public JNICallTrampolineMethod getCallTrampolineMethod(CallVariant variant, boolean nonVirtual) {
-        JNICallTrampolineMethod method = null;
-        if (variant == CallVariant.VARARGS) {
-            method = nonVirtual ? varargsNonvirtualCallTrampolineMethod : varargsCallTrampolineMethod;
-        } else if (variant == CallVariant.ARRAY) {
-            method = nonVirtual ? arrayNonvirtualCallTrampolineMethod : arrayCallTrampolineMethod;
-        } else if (variant == CallVariant.VA_LIST) {
-            method = nonVirtual ? valistNonvirtualCallTrampolineMethod : valistCallTrampolineMethod;
+        String trampolineName = JNIJavaCallTrampolines.getTrampolineName(variant, nonVirtual);
+        return getCallTrampolineMethod(trampolineName);
+    }
+
+    public JNICallTrampolineMethod getCallTrampolineMethod(String javaTrampolineName) {
+        JNICallTrampolineMethod jniCallTrampolineMethod = trampolineMethods.get(javaTrampolineName);
+        assert jniCallTrampolineMethod != null;
+        return jniCallTrampolineMethod;
+    }
+
+    public JNICallTrampolineMethod getOrCreateCallTrampolineMethod(DuringSetupAccessImpl access, String trampolineName) {
+        JNICallTrampolineMethod jniCallTrampolineMethod = trampolineMethods.get(trampolineName);
+
+        if (jniCallTrampolineMethod != null) {
+            return jniCallTrampolineMethod;
         }
-        assert method != null;
-        return method;
+
+        MetaAccessProvider wrappedMetaAccess = access.getMetaAccess().getWrapped();
```

Comment:
It's correct to use the wrapped meta access here, I suppose that's where you ran into trouble with the `AnalysisMetaAccess`, right?

Reply:
I have trouble in both cases if I don't use the wrapped meta access.